File: dashdown/static/components/daterange.js
// Dashdown DateRange Component
// Date range picker with preset ranges and URL sync

"use strict";

import { parseUrlParams } from "../core.js";

/**
 * Sync date range to URL
 * @param {string} startParam - Start date parameter name
 * @param {string} endParam - End date parameter name
 * @param {string} startDate - Start date value
 * @param {string} endDate - End date value
 */
export function syncDateRangeToUrl(startParam, endParam, startDate, endDate) {
  const params = new URLSearchParams(window.location.search);

  if (startDate) {
    params.set(startParam, startDate);
  } else {
    params.delete(startParam);
  }

  if (endDate) {
    params.set(endParam, endDate);
  } else {
    params.delete(endParam);
  }

  const newUrl = `${window.location.pathname}?${params.toString()}`;
  window.history.pushState({}, "", newUrl);

  // Trigger custom event for other components
  window.dispatchEvent(new CustomEvent("dashdown:url-updated", {
    detail: { params: Object.fromEntries(params.entries()) }
  }));
}

/**
 * Format a Date as YYYY-MM-DD in local time (not UTC, unlike toISOString).
 * @param {Date} d - Date to format
 * @returns {string} - Local date string
 */
function formatLocalDate(d) {
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, "0");
  const day = String(d.getDate()).padStart(2, "0");
  return `${y}-${m}-${day}`;
}

/**
 * Resolve a preset config to a concrete {start, end} date pair, or null for
 * presets that carry no fixed range (custom / unknown). Single source of preset
 * math — both setPreset (apply) and updateActivePreset (detect) call this so the
 * two can never drift. All arithmetic is in LOCAL time to match formatLocalDate.
 * @param {object} preset - A preset config from preset_configs
 * @param {Date} today - Local midnight of the current day
 * @returns {{start: Date, end: Date}|null}
 */
function presetRange(preset, today) {
  if (!preset) return null;
  if (preset.kind === "rolling") {
    const start = new Date(today);
    start.setDate(today.getDate() + preset.days_start);
    const end = new Date(today);
    end.setDate(today.getDate() + preset.days_end);
    return { start, end };
  }
  if (preset.kind === "calendar") {
    return calendarRange(preset.unit, preset.offset, today);
  }
  return null;
}

/**
 * Compute a calendar-aligned range. offset 0 = current period start → today
 * (period-to-date, so it never selects future days with no data); offset -1 =
 * the whole previous period (its start → its end). Weeks start Monday (ISO).
 * @param {string} unit - "week" | "month" | "year"
 * @param {number} offset - 0 = current period, -1 = previous period
 * @param {Date} today - Local midnight of the current day
 * @returns {{start: Date, end: Date}|null}
 */
function calendarRange(unit, offset, today) {
  if (unit === "week") {
    // getDay(): 0=Sun..6=Sat. Days elapsed since Monday (0=Mon..6=Sun):
    const sinceMonday = (today.getDay() + 6) % 7;
    const start = new Date(today);
    start.setDate(today.getDate() - sinceMonday + offset * 7);
    if (offset === 0) return { start, end: new Date(today) };
    const end = new Date(start);
    end.setDate(start.getDate() + 6);
    return { start, end };
  }
  if (unit === "month") {
    const start = new Date(today.getFullYear(), today.getMonth() + offset, 1);
    if (offset === 0) return { start, end: new Date(today) };
    // Day 0 of the next month = last day of the target month.
    const end = new Date(today.getFullYear(), today.getMonth() + offset + 1, 0);
    return { start, end };
  }
  if (unit === "year") {
    const start = new Date(today.getFullYear() + offset, 0, 1);
    if (offset === 0) return { start, end: new Date(today) };
    const end = new Date(today.getFullYear() + offset, 11, 31);
    return { start, end };
  }
  return null;
}

// Define the Alpine component function globally so it can be used in HTML
// This will be called when Alpine initializes the component
window.dateRangeComponent = function(name) {
  return {
    startDate: "",
    endDate: "",
    activePreset: "",
    presets: {},
    persist: false,
    defaultPreset: "",
    storageKey: "",

    init() {
      // Read config from data-config attribute (properly HTML-escaped JSON)
      let config = {};
      try {
        config = JSON.parse(this.$el.dataset.config || "{}");
      } catch (e) {
        console.error("DateRange: failed to parse data-config", e);
      }

      // Load preset configs from config
      this.presets = config.preset_configs || {};

      const startParam = config.start_param || `${name}_start`;
      const endParam = config.end_param || `${name}_end`;

      // Persistence (localStorage) + default preset power the project-wide global
      // date filter: remember the selection across navigation, and seed it on the
      // first ever visit. Keyed by the param pair so every page sharing the same
      // global filter reads the same value.
      this.persist = !!config.persist;
      this.defaultPreset = config.default || "";
      this.storageKey = `dashdown:date-filter:${startParam}|${endParam}`;

      // Load from URL on initialization
      const urlParams = parseUrlParams();
      let seededFromUrl = false;
      if (urlParams[startParam]) {
        this.startDate = urlParams[startParam];
        seededFromUrl = true;
      }
      if (urlParams[endParam]) {
        this.endDate = urlParams[endParam];
        seededFromUrl = true;
      }

      // Mark both date keys as visible controls so they aren't shown as chips.
      if (window.Alpine && Alpine.store && Alpine.store("filterControls")) {
        Alpine.store("filterControls")[startParam] = true;
        Alpine.store("filterControls")[endParam] = true;
      }

      this.updateActivePreset();
      this.setupUrlSync(startParam, endParam);

      // URL wins over the remembered/default range (a shared deep link must
      // override). With nothing in the URL: restore the persisted value, else
      // apply the default preset. Both set the dates *after* setupUrlSync, so the
      // watchers push them to the filters store + URL (the single reactive path).
      if (!seededFromUrl) {
        const stored = this.persist ? this.readStored() : null;
        if (stored && (stored.start || stored.end)) {
          this.startDate = stored.start || "";
          this.endDate = stored.end || "";
          this.updateActivePreset();
        } else if (this.defaultPreset) {
          this.setPreset(this.defaultPreset);
        }
      } else if (this.persist) {
        // Mirror a URL-provided range into storage so other pages inherit it.
        this.writeStored();
      }
    },

    readStored() {
      try {
        return JSON.parse(window.localStorage.getItem(this.storageKey) || "null");
      } catch (e) {
        return null;
      }
    },

    writeStored() {
      if (!this.persist) return;
      try {
        if (this.startDate || this.endDate) {
          window.localStorage.setItem(
            this.storageKey,
            JSON.stringify({ start: this.startDate, end: this.endDate })
          );
        } else {
          window.localStorage.removeItem(this.storageKey);
        }
      } catch (e) {
        /* storage disabled / over quota — persistence is best-effort */
      }
    },

    setupUrlSync(startParam, endParam) {
      const urlSync = this.$el.dataset.urlSync === "true";

      const pushToFilters = () => {
        // Mutate the existing store keys rather than replacing the whole store
        // object. Replacing it re-triggers every other filter component's
        // reactive URL-sync effect, which re-dispatches stale `url-updated`
        // events that would clobber the dates we just set (single-path rule).
        const store = window.Alpine && Alpine.store ? Alpine.store("filters") : null;
        if (store) {
          store[startParam] = this.startDate || "";
          store[endParam] = this.endDate || "";
        }
        // Persist after each change so the global filter survives navigation.
        this.writeStored();
        // Writing the store keys above is the single reactive path: data
        // components subscribe to `$store.filters` via Alpine effects and
        // re-render on change. No custom broadcast event needed.
      };

      // Watch for changes and sync to filters store (and optionally URL)
      this.$watch("startDate", () => {
        pushToFilters();
        if (urlSync) syncDateRangeToUrl(startParam, endParam, this.startDate, this.endDate);
      });

      this.$watch("endDate", () => {
        pushToFilters();
        if (urlSync) syncDateRangeToUrl(startParam, endParam, this.startDate, this.endDate);
      });

      if (!urlSync) return;

      // Handle popstate events
      window.addEventListener("popstate", () => {
        const urlParams = parseUrlParams();
        if (urlParams[startParam] !== this.startDate) {
          this.startDate = urlParams[startParam] || "";
        }
        if (urlParams[endParam] !== this.endDate) {
          this.endDate = urlParams[endParam] || "";
        }
        this.updateActivePreset();
      });

      // NOTE: no `dashdown:url-updated` listener here. Only this component ever
      // writes the date params, so reacting to that event is a pure feedback
      // loop — stale events from other filters would overwrite a freshly-picked
      // preset. Browser back/forward is handled by popstate above.
    },

    setPreset(presetName) {
      // The empty "All time" option clears the range entirely (the watchers
      // push the empty dates to the store/URL like any other change).
      if (!presetName) {
        this.startDate = "";
        this.endDate = "";
        this.activePreset = "";
        return;
      }
      const preset = this.presets[presetName];
      if (!preset) return;

      // "Custom" carries no fixed range: just reveal the inputs and keep
      // whatever dates are already set so the user can adjust them.
      if (preset.kind === "custom") {
        this.activePreset = "custom";
        return;
      }

      const today = new Date();
      today.setHours(0, 0, 0, 0);

      const range = presetRange(preset, today);
      if (range) {
        // Format in LOCAL time. toISOString() converts to UTC, which rolls the
        // day back/forward in non-UTC timezones — that desynced these dates from
        // updateActivePreset()'s local-time comparison, so the active preset
        // never matched and the day was off by one.
        this.startDate = formatLocalDate(range.start);
        this.endDate = formatLocalDate(range.end);
        this.activePreset = presetName;
      }
    },

    updateFromInputs() {
      this.activePreset = "custom";
    },

    updateActivePreset() {
      if (!this.startDate && !this.endDate) {
        this.activePreset = "";
        return;
      }
      // A half-set range (deep link with only one param) is "custom" so the
      // date inputs stay visible rather than the select claiming "All time".
      if (!this.startDate || !this.endDate) {
        this.activePreset = "custom";
        return;
      }

      const today = new Date();
      today.setHours(0, 0, 0, 0);

      const start = new Date(this.startDate + "T00:00:00");
      const end = new Date(this.endDate + "T00:00:00");

      for (const [presetName, preset] of Object.entries(this.presets)) {
        const range = presetRange(preset, today);
        if (
          range &&
          start.getTime() === range.start.getTime() &&
          end.getTime() === range.end.getTime()
        ) {
          this.activePreset = presetName;
          return;
        }
      }

      this.activePreset = "custom";
    }
  };
};

File: dashdown/components/builtin/date_range.py
from __future__ import annotations

from dashdown.components.base import Component, RenderContext, register_component
from dashdown.components.builtin._util import (
    attr_str,
    esc,
    filter_bar_marker,
    new_id,
    resolve_debounce,
    safe_json,
)


# Preset date ranges. Two kinds the frontend knows how to resolve:
#   - "rolling": a window measured in days back from today. days_start/days_end
#     are offsets (0 = today, -6 = six days ago).
#   - "calendar": anchored to a calendar unit (week/month/year). offset 0 = the
#     current period from its start through today (month/week/year-to-date);
#     offset -1 = the whole previous period (its start through its end).
# Calendar ranges can't be expressed as fixed day-offsets (months/years vary in
# length), so daterange.js computes the boundaries from the viewer's local
# "today" — see presetRange()/calendarRange() there. Weeks start Monday (ISO).
PRESET_RANGES = {
    "today": {"label": "Today", "kind": "rolling", "days_start": 0, "days_end": 0},
    "yesterday": {"label": "Yesterday", "kind": "rolling", "days_start": -1, "days_end": -1},
    "last_7_days": {"label": "Last 7 Days", "kind": "rolling", "days_start": -6, "days_end": 0},
    "last_30_days": {"label": "Last 30 Days", "kind": "rolling", "days_start": -29, "days_end": 0},
    "last_90_days": {"label": "Last 90 Days", "kind": "rolling", "days_start": -89, "days_end": 0},
    "this_week": {"label": "This Week", "kind": "calendar", "unit": "week", "offset": 0},
    "last_week": {"label": "Last Week", "kind": "calendar", "unit": "week", "offset": -1},
    "this_month": {"label": "This Month", "kind": "calendar", "unit": "month", "offset": 0},
    "last_month": {"label": "Last Month", "kind": "calendar", "unit": "month", "offset": -1},
    "this_year": {"label": "This Year", "kind": "calendar", "unit": "year", "offset": 0},
    "last_year": {"label": "Last Year", "kind": "calendar", "unit": "year", "offset": -1},
    "custom": {"label": "Custom", "kind": "custom"},
}


@register_component("DateRange")
class DateRange(Component):
    """Date range picker component with preset ranges and custom selection.

    Renders as a compact pill: inline label + a preset <select> ("All time"
    clears the range); picking "Custom" reveals start/end date inputs.

    Supports URL sync by default (url_sync=True).
    Selected date range is stored in the page-level Alpine store under `filters[name]`.

    Parameters:
    - name: Required. The filter name to store the date range under.
    - label: Optional. Display label (defaults to name; shown as the inline prefix).
    - presets: Optional. Comma-separated list of preset ranges, in display order
      (default: today,last_7_days,last_30_days,last_90_days,custom). Available:
      today, yesterday, last_7_days, last_30_days, last_90_days, this_week,
      last_week, this_month, last_month, this_year, last_year, custom. The
      calendar presets (this_/last_ week/month/year) are anchored to real
      calendar boundaries, not day-offset approximations.
    - start_param: Optional. URL parameter name for start date (defaults to name + "_start")
    - end_param: Optional. URL parameter name for end date (defaults to name + "_end")
    - url_sync: Optional. Enable URL sync (default: True)
    - default: Optional. A preset name (e.g. "last_30_days") applied on first load
      when nothing is set (no URL params and no persisted value).
    - persist: Optional. Remember the selection in localStorage so it survives
      navigation — the mechanism behind the project-wide global date filter.
    - debounce: Optional. Quiet period (ms) before a change re-fetches data, so a
      preset (which sets start+end) or a custom edit coalesces into one fetch.
      Defaults to the project-wide ``filters.debounce`` (300 unless raised).
    - bar: Optional. Relocate this control into the page's top filter bar.
      Default is inline (renders where authored).

    Example:
    <DateRange name="date_range" label="Date Range" presets="today,last_7_days,last_30_days,custom" />
    
    URL format:
    ?date_range_start=2024-01-01&date_range_end=2024-01-31
    """

    is_filter = True

    def render(
        self, attrs, ctx: RenderContext, inner: str | None = None
    ) -> str:
        name = attr_str(attrs, "name")
        if not name:
            raise ValueError("DateRange requires a `name` attribute")
        
        label = attr_str(attrs, "label", name)
        presets_str = attr_str(attrs, "presets", "today,last_7_days,last_30_days,last_90_days,custom")
        start_param = attr_str(attrs, "start_param", f"{name}_start")
        end_param = attr_str(attrs, "end_param", f"{name}_end")
        url_sync = attrs.get("url_sync", True)
        # `default="last_30_days"` applies a preset on first load when nothing is
        # set (no URL params, no persisted value). `persist` remembers the
        # selection in localStorage so it survives navigation — the basis of the
        # project-wide global date filter (see GlobalDateFilterConfig).
        default_preset = attr_str(attrs, "default", "") or None
        persist = bool(attrs.get("persist", False))
        # Debounce the store write (data re-fetch): a preset sets start+end and a
        # custom edit touches each input, so coalesce them into one fetch. The
        # initial seed stays immediate (daterange.js). Per-control `debounce=`
        # wins, else the project-wide `filters.debounce` (see resolve_debounce).
        debounce = resolve_debounce(attrs, ctx)
        # Inline by default (renders where authored); `bar` relocates it into the
        # top filter row (read by filter_bar.js). The global date control passes
        # `filter_bar=True` in embed mode to route itself there. See filter_bar_marker.
        filter_bar_attr = filter_bar_marker(attrs, ctx)

        # Parse presets
        presets = [p.strip() for p in presets_str.split(",") if p.strip()]

        cid = new_id("dashdown-daterange")

        # Presets render as <option>s in a compact select; the empty "All time"
        # option clears the range (handled by setPreset).
        preset_options = ['<option value="">All time</option>']
        preset_options += [
            f'<option value="{esc(p)}">{esc(PRESET_RANGES[p]["label"])}</option>'
            for p in presets
            if p in PRESET_RANGES
        ]
        preset_html = "".join(preset_options)

        # Build full preset configs for selected presets only
        preset_configs = {p: PRESET_RANGES[p] for p in presets if p in PRESET_RANGES}

        # Build config for frontend
        config = {
            "name": name,
            "label": label,
            "presets": presets,
            "preset_configs": preset_configs,
            "start_param": start_param,
            "end_param": end_param,
            "url_sync": url_sync,
            "default": default_preset,
            "persist": persist,
            "debounce": debounce,
        }

        # Compact pill: inline label prefix + preset select; the two date inputs
        # only appear when the "Custom" preset is active so
        # the common case stays one small control. Labels move to aria-labels.
        return (
            f'<div class="dashdown-daterange dashdown-filter-pill" id="{cid}" '
            f'data-async-component="daterange" '
            f'data-config="{esc(safe_json(config))}" '
            f'data-name="{esc(name)}" '
            f'data-start-param="{esc(start_param)}" '
            f'data-end-param="{esc(end_param)}" '
            f'data-url-sync="{str(url_sync).lower()}" '
            f"{filter_bar_attr}"
            f'x-data="dateRangeComponent(\'{esc(name)}\')" '
            f">"
            f'<span class="dashdown-filter-pill-label">{esc(label)}'
f'<span class="dashdown-filter-pill-colon">:</span></span>'
            f'<select class="select select-sm" aria-label="{esc(label)} preset" '
            f'x-model="activePreset" '
            f'@change="setPreset($event.target.value)" '
            f'>{preset_html}</select>'
            f'<div class="dashdown-daterange-custom" '
            f'x-show="activePreset === \'custom\'" x-cloak '
            f'>'
            f'<input type="date" class="input input-sm" '
            f'x-model="startDate" '
            f'x-bind:max="endDate" '
            f'@change="updateFromInputs" '
            f'aria-label="{esc(label)} start date" '
            f'>'
            f'<span class="text-base-content/60">–</span>'
            f'<input type="date" class="input input-sm" '
            f'x-model="endDate" '
            f'x-bind:min="startDate" '
            f'@change="updateFromInputs" '
            f'aria-label="{esc(label)} end date" '
            f'>'
            f'</div>'
            f'</div>'
        )
